### PR TITLE
TF importer: Corrected end tensor_content parsing for StridedSlice layer.

### LINKED ISSUE #16337: TF importer: StridedSlice gives different shape compared to TF
##### System information (version)
- OpenCV == 4.2.0-dev
- python == 3.6
- Operating System / Platform == Ubuntu 18.04 x86_64
- compiler == gcc 7.4.0
- tensorflow == 1.14.0

##### Detailed description
Using a CNN model from tensorflow that contains a `Cropping2D` layer (or another layer that leads to a `StridedSlice`) does not give the correct output shapes in OpenCV:
Shape in OpenCV is: `(4, 8, 32, 32)`
Shape in Tensorflow is: `(4, 31, 31, 8)`
Problem seems to be related to [tf_importer.cpp#L1472](https://github.com/opencv/opencv/blob/fb61f88b9c32c22c222a55f885c4b6a7f475b396/modules/dnn/src/tensorflow/tf_importer.cpp#L1472), which treats unchanged dimensions the same as one pixel cutoffs.

##### Steps to reproduce
The following python script can be used to trigger the bug:

  ```.py
#!/usr/bin/env python3
import os

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = ""
import numpy
import cv2
import tensorflow as tf
from tensorflow.keras.layers import Cropping2D, Input, Conv2D
from tensorflow.tools.graph_transforms import TransformGraph

if __name__ == '__main__':
    print(cv2.__version__)
    batch_size = 4
    input_shape = [batch_size, 32, 32, 2]
    features = Input(input_shape[1:], name='input')
    features0 = Conv2D(filters=8, kernel_size=[3, 3], padding='same')(features)
    features0 = Cropping2D(((0, 1), (0, 1)))(features0)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        res = sess.run(features0, {features: numpy.zeros(input_shape)})
        print(res.shape)
        constant_graph = tf.graph_util.convert_variables_to_constants(sess, sess.graph.as_graph_def(),
                                                                      ['cropping2d/strided_slice'])
        tf.train.write_graph(constant_graph, "", "graph_final.pb", as_text=False)

    # export
    with tf.gfile.FastGFile("graph_final.pb", 'rb') as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())
        graph_def = TransformGraph(graph_def, ['input'], ['cropping2d/strided_slice'],
                                   ['strip_unused_nodes', 'fold_constants'])
        with tf.gfile.FastGFile('saved_model.pb', 'wb') as f:
            f.write(graph_def.SerializeToString())

    # read model
    cvNet = cv2.dnn.readNetFromTensorflow('./saved_model.pb')
    img = numpy.zeros((batch_size, 2, 32, 32), dtype='float32')
    cvNet.setInput(img)
    cvOut = cvNet.forward()
    print(cvOut.shape)
 ```

### PR DESCRIPTION
**Pull request by ashishkrshrivastava**
**Merge with extra**: [opencv/opencv_extra#704](https://github.com/opencv/opencv_extra/pull/704)
resolves #16337

### This pullrequest changes 
Setting of end Mat values using tensor_content of "cropping2d/strided_slice/stack_1" from pbtxt file. 

`opencv_extra=opencv_extratf`

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r3.0:16.04
build_image:Custom Win=openvino-2019r3.0
build_image:Custom Mac=openvino-2019r3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 1]
```diff
@@ -1468,6 +1468,8 @@ void TFImporter::populateNet(Net dstNet)
             int end_mask = getLayerAttr(layer, "end_mask").i();
             for (int i = 0; i < num; ++i)
             {
+                if (ends.at<int>(i) <= 0)
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_importer.cpp]: Please add a test for this

Comment (ashishkrshrivastava) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: @dkurt , I have added the test, model and data.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 2]
```diff
@@ -1468,6 +1468,8 @@ void TFImporter::populateNet(Net dstNet)
...
+                    ends.at<int>(i) -= 1;
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_importer.cpp]: According to https://www.tensorflow.org/api_docs/python/tf/strided_slice:

> If the ith bit of begin_mask is set, begin[i] is ignored and the fullest possible range in that dimension is used instead. end_mask works analogously, except with the end range.

Can you please show the values of `ends` for new test?

Comment (ashishkrshrivastava) [reply modules/dnn/src/tensorflow/tf_importer.cpp]:
```
attr {
    key: "end_mask"
    value {
      i: 9
    }
  }

attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
          dim {
            size: 4
          }
        }
        tensor_content: "\000\000\000\000\377\377\377\377\377\377\377\377\000\000\000\000"
      }
    }
  }
```
so end mask = 1 0 0 1 (i.e. 9)
and end is [0, -1, -1, 0]

Comment (dkurt) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: Ok, got it. That means ends are excluded from range: `[begin, end)`. Thanks!

So it's better to change condition to `if (ends.at<int>(i) < 0)`. Zero value should not be processed.

**Comment by dkurt** [comment]
Please increase test tolerance for Myriad device:
```
[ RUN      ] Test_TensorFlow_layers.slice/1, where GetParam() = DLIE/MYRIAD
/build/precommit_custom_linux/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normL1) <= (l1), actual: 0.00482579 vs 0.004
[  FAILED  ] Test_TensorFlow_layers.slice/1, where GetParam() = DLIE/MYRIAD (76 ms)
```
https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/3311

**Comment by ashishkrshrivastava** [comment]
>  Well done! Thanks for your contribution!

Thanks.
